Add retry logic for initial parameters in ExternalSampler (#2739)

0AnshuAditya0 opened the pull request:
Fixes #2739

This PR adds retry logic when generating initial parameters for external samplers, similar to what HMC already does.

**Changes:**
- Modified `src/mcmc/external_sampler.jl` to attempt finding valid initial parameters up to 10 times
- Checks that both logp and gradient are finite before accepting initial parameters
- Provides informative error message if all attempts fail

**Why:**
External samplers previously only tried once to generate initial parameters. If the first attempt produced infinite logp or gradient, sampling would fail immediately. This makes external samplers more robust.

penelopeysm commented on test/mcmc/abstractmcmc.jl:
```diff
@@ -150,4 +150,129 @@ end
+            @test size(chain, 1) == 10
+            x_samples = chain[:x]
+            @test all(abs.(x_samples) .< 0.1)
+        end
```
As I said before, this test is not doing anything that the previous one didn't do.

0AnshuAditya0 commented:
i reverted the commit while trying to simplify the tests, but I also reverted some valid changes by mistake, especially the HMC-specific test and the cleaned validator logic 
that wasnt intentional
i just accidently did model() instead of bad_int_model_hmc() or there is something else i did i cant figure out

penelopeysm commented:
The test in the earlier commit was fine before you reverted it. No don't use abs(x) please, the Ref counter was exactly what's needed.

penelopeysm reviewed (changes requested):
There are still other things from my previous review which were in the reverted commit.
